Add a modern CSS spinner

Jaifroid opened the pull request:
@mossroy I've taken the CSS spinner code as far as I can without hooking into the Service Worker messaging channel. So it works fully in jQuery mode, and in a limited way in Service Worker mode: when loading a new page by searching for it, with the Random button, or when loading the landing page, but not when clicking on a link within a page. The latter would require hooking into the messaging channel, and so would require a policy decision if we want uniformity across the app. It can be left for future decision.

Regarding your comment in #425 that keeping the old page on-screen until we have the new html to replace it might cause the app to keep requesting images unnecessarily, I have done extensive testing in #426 : there is a broader issue affecting master, not this PR specifically, so I have opened a separate issue for it. 

Could you please check this code and test? I'd be grateful if you could test in particular that #361  ("TypeError: can't access dead objects") is not reintroduced. It shouldn't be, as we still have the specific fix for that in place. There is also a related query embedded in a comment in app.js line 748 addressed to you, which I'll remove when the query is resolved. No rush at all!

Jaifroid commented:
Thanks @mossroy : 

- I'll experiment with a darker colour -- I agree it's sometimes hard to notice;
- I was having trouble finding a good place for "Caching styles..."; it's not so necessary now that xzdec is much faster, though could still be useful on low-end devices. I'll experiment with what it looks like placing it over the spinner, otherwise use more noticeable style;
- Good catch about the scrolling -- I think it's not necessary, will check;
- I think the only purpose of `$('#articleContent').contents().remove();` would be to tell jQuery to remove its hooks, so I think in SW mode it only needs to be performed once when the user switches to SW mode after having used jQuery mode -- I'll look for a place to relocate it in the switching code.

Jaifroid commented:
Hi @mossroy , if you get a moment, could you give me an opinion on the (visual) implementation now?

- More visible foreground and background colours for spinner;
- In jQuery mode, Caching styles... message now centred beneath spinner;
- Caching styles... message is updated with abbreviated name of each css file as it is extracted.

Note that the caching message(s) will only show in jQuery mode, as the cache isn't implemented for SW mode. Also the caching message will only be displayed on first load of the landing page, and very rarely on a page that needs a new CSS file that has not yet been cached.

Specifically, does the caching message strike a good balance between being discreet but still being noticeable/informative? Is it too obtrusive? Would it be better relocated to bottom-right or top-right corners? (We can't use bottom left because many browsers use that area to show the URL of a link that has been clicked.) 

An example over blank background shown below.

![image](https://user-images.githubusercontent.com/4304337/46862913-e7950580-ce0d-11e8-8acd-5ba7b654fa69.png)

Jaifroid commented:
Great, thanks! I can certainly test adding a line to the onclick code to activate the spinner sooner for these edge cases. I think it can be part of this PR.

I've still got code clearing-up to do and one small thing to fix from earlier discussion. I'll ask you to look over final code when it's ready (probably at weekend) before merging.

Jaifroid commented:
@mossroy I've completed the pending issues (see commit titles). Could you test your slow I/O files with this final code? The spinner should display on click of UI elements in both modes, and inline article links in jQuery mode.

If you get a moment to look through the code, it would be great. Quite a lot of redundant hiding and showing has been deleted.

There is one thing that might need decision: I got rid of the "Reading article xxxxx ...." message before I implemented the css caching info box beneath the spinner, on the basis that users know what they've clicked on, and the spinner is enough feedback. However, it would now be possible to use the css caching info box to present that message. I tend to think it's still not necessary, and displaying the message on each page load would mean that we would always see the box. Currently it is only shown when there is new CSS to cache, so only for the first couple of page loads usually. I feel it's neater just to use the spinner as feedback. The "loss" of feedback is mitigated by the fact that the article list stays in place now until the page is ready, and when a user clicks on an inline link, browsers usually display the link in their status message bottom left.

However, I thought I should point this out, since it is removing a feature that exists in master.

mossroy commented:
I tested with the same slow I/Os (on Firefox 60ESR and IE11, both in jQuery mode) : the user feedback is much better now, with the spinner appearing right after the click.
A small regression : the spinner does not seem to appear any more when clicking on "Random" button.

Regarding the "loss" of "Reading article xxx", I don't have a strong opinion. We can probably leave it this way : it's still possible to add it back later if we feel it's necessary. In a sense, it's also more consistent with the SW mode.

I had a quick look at the code : I like to see some code removal in this hiding/showing code :-), see #78.

I did not test in SW mode, but if you did, you might squash/rebase/merge this PR.

Jaifroid commented:
Thanks for spotting the regression. It was a bit "random", and a puzzle, because I can't see the rogue `hide()` statement that turns off the spinner. It's clearly shown at the very beginning of the `goToRandomArticle()` function, and nothing in the chain should turn it off until the html is received and all CSS is loaded. I wonder if there is a manifestation of #426 here: a delayed CSS file from a previous page comes through async and triggers turning off the spinner (because we have a tendency when testing to mash the Random button (at least I do), and because I moved the `show()` statement to before the dirEntry is sourced, it gets turned off before we read the file. In any case, I've fixed it with an extra `show()` statement. But it feels like a brute force kludge... Hmmm.

Jaifroid commented:
Found another regression in this PR with FFOS, when clicking on a search result, we get a severely truncated display area once the new page is shown, at least in the simulator. It's to do with display of the on-screen keyboard, I guess. I need to remove the article list before that code is called...

It's never easy!!

Jaifroid commented:
OK, [this commit](https://github.com/kiwix/kiwix-js/pull/427/commits/a363a93ea9644c068901c8368ae3fdf9968f8bef) solves that one. Again, feels a bit like brute force, but it does make senes to call `resizeIFrame()` after we render CSS.

I'm going to leave this till tomorrow before squashing etc. @mossroy , I don't know if you want to test on a real FFOS device first? If not, it does seem to work in the simulator. Anyway, I aim to merge early tomorrow morning unless I hear to the contrary.
